File: src/main/java/com/gradecalc/util/ExcelExporter.java
package com.gradecalc.util;

import com.gradecalc.model.GradeCalculator;
import com.gradecalc.model.Subject;
import org.apache.poi.xssf.usermodel.*;
import java.io.FileOutputStream;

public class ExcelExporter {

    public static void export(GradeCalculator calc, String filePath)
            throws Exception {
        XSSFWorkbook workbook = new XSSFWorkbook();
        XSSFSheet sheet = workbook.createSheet("Grade Report");

        // Header row
        String[] headers = {"Subject", "Score", "Max",
                "Percentage", "Grade", "Status"};
        XSSFRow header = sheet.createRow(0);
        for (int i = 0; i < headers.length; i++) {
            header.createCell(i).setCellValue(headers[i]);
        }

        // Data rows — lambda with index
        var subjects = calc.getSubjects();
        for (int i = 0; i < subjects.size(); i++) {
            Subject s = subjects.get(i);
            double pct = s.getPercentage();
            XSSFRow row = sheet.createRow(i + 1);
            row.createCell(0).setCellValue(s.getName());
            row.createCell(1).setCellValue(s.getScore());
            row.createCell(2).setCellValue(s.getMaxScore());
            row.createCell(3).setCellValue(
                    String.format("%.1f%%", pct));
            row.createCell(4).setCellValue(
                    calc.getGradeLetter(pct));
            row.createCell(5).setCellValue(
                    calc.getPassFail(pct));
        }

        // Summary rows
        int sumRow = subjects.size() + 2;
        sheet.createRow(sumRow).createCell(0)
                .setCellValue("Average: " +
                        String.format("%.1f%%", calc.calculate()));
        sheet.createRow(sumRow + 1).createCell(0)
                .setCellValue("GPA: " + calc.calculateGPA());
        sheet.createRow(sumRow + 2).createCell(0)
                .setCellValue("Status: " +
                        calc.getOverallPassFail());

        try (FileOutputStream out = new FileOutputStream(filePath)) {
            workbook.write(out);
        }
        workbook.close();
    }
}
```

        ---

        ## STEP 7 — Design the GUI in Scene Builder

Open **Scene Builder** and build this layout, then save it as `main-view.fxml`:
        ```
        ┌─────────────────────────────────────────┐
        │         🎓 Student Grade Calculator      │
        ├──────────┬──────────┬────────────────────┤
        │ Subject  │  Score   │  Max Score  [Add]  │
        ├──────────┴──────────┴────────────────────┤
        │  TableView (Subject | Score | Grade |    │
        │             Status | Actions[Delete])    │
        ├─────────────────────────────────────────┤
        │  Average: __  GPA: __  Status: __       │
        ├─────────────────────────────────────────┤
        │  [Delete All]  [Download PDF] [Excel]   │
        └─────────────────────────────────────────┘

File: student-grade-calculator-android/src/main/kotlin/exporters/ExcelExporter.kt
package exporters

import models.Student
import org.apache.poi.ss.usermodel.FillPatternType
import org.apache.poi.ss.usermodel.IndexedColors
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.ByteArrayOutputStream
import java.util.Base64

/**
 * Exports student results as an Excel (.xlsx) workbook.
 *
 * POLYMORPHISM: Implements [ReportExporter]. Uses Apache POI internally
 * but exposes exactly the same interface as Html and Json exporters.
 *
 * Because Excel requires binary output (not a plain String), this exporter
 * returns a Base64-encoded string of the workbook bytes. The [FileManager]
 * knows to decode it back when writing to disk.
 *
 * No loops — uses [forEachIndexed] everywhere instead of for/while.
 */
class ExcelExporter : ReportExporter {

    override fun fileExtension(): String = "xlsx"
    override fun formatName(): String    = "Excel"

    /**
     * Builds an XLSX workbook and returns its content as a Base64 string.
     *
     * Layout:
     * | StudentID | Name | Type | Subject… | Average | Grade | GPA | Status |
     */
    override fun export(students: List<Student>, subjectHeaders: List<String>): String {
        val workbook = XSSFWorkbook()
        val sheet    = workbook.createSheet("Results")

        // ── Styles ──────────────────────────────────────────────────────────
        val headerStyle = workbook.createCellStyle().apply {
            fillForegroundColor = IndexedColors.DARK_BLUE.index
            fillPattern         = FillPatternType.SOLID_FOREGROUND
            val font = workbook.createFont().also {
                it.color = IndexedColors.WHITE.index
                it.bold  = true
            }
            setFont(font)
        }
        val passStyle = workbook.createCellStyle().apply {
            fillForegroundColor = IndexedColors.LIGHT_GREEN.index
            fillPattern         = FillPatternType.SOLID_FOREGROUND
        }
        val failStyle = workbook.createCellStyle().apply {
            fillForegroundColor = IndexedColors.ROSE.index
            fillPattern         = FillPatternType.SOLID_FOREGROUND
        }

        // ── Header row ───────────────────────────────────────────────────────
        val headers = listOf("StudentID", "Name", "Type") +
                subjectHeaders +
                listOf("Average", "Grade", "GPA", "Status")

        val headerRow = sheet.createRow(0)
        headers.forEachIndexed { col, title ->
            headerRow.createCell(col).also {
                it.setCellValue(title)
                it.cellStyle = headerStyle
            }
        }

        // ── Data rows ────────────────────────────────────────────────────────
        students.forEachIndexed { rowIdx, student ->
            val row      = sheet.createRow(rowIdx + 1)
            val rowStyle = if (student.grade.status() == "PASS") passStyle else failStyle

            row.createCell(0).also { it.setCellValue(student.studentId); it.cellStyle = rowStyle }
            row.createCell(1).also { it.setCellValue(student.name);      it.cellStyle = rowStyle }
            row.createCell(2).also { it.setCellValue(student.studentType()); it.cellStyle = rowStyle }

            student.scores.forEachIndexed { i, score ->
                row.createCell(3 + i).also { it.setCellValue(score); it.cellStyle = rowStyle }
            }

            val after = 3 + student.scores.size
            row.createCell(after    ).also { it.setCellValue(student.average); it.cellStyle = rowStyle }
            row.createCell(after + 1).also { it.setCellValue(student.grade.letter);   it.cellStyle = rowStyle }
            row.createCell(after + 2).also { it.setCellValue(student.grade.gpa);     it.cellStyle = rowStyle }
            row.createCell(after + 3).also { it.setCellValue(student.grade.status());  it.cellStyle = rowStyle }
        }

        // Auto-size all columns for readability
        (0 until headers.size).forEach { sheet.autoSizeColumn(it) }

        // ── Encode as Base64 so we can return a String ───────────────────────
        val bytes = ByteArrayOutputStream()
            .also { workbook.write(it) }
            .toByteArray()
        workbook.close()

        return Base64.getEncoder().encodeToString(bytes)
    }
}